domains/client/domain-operator/src/operator.rs

```diff
@@ -250,9 +254,60 @@ where
             spawn_essential.spawn_essential("domain-sync", None, Box::pin(domain_sync_task));
         }
 
-        spawn_essential.spawn_essential_blocking(
-            "domain-operator-worker",
-            None,
+        let start_worker_task = {
+            // Wait for the target block importing if we snap syncing
+            if let Some(ref snap_sync_orchestrator) = snap_sync_orchestrator {
+                let target_block_number = snap_sync_orchestrator.target_block().await;
+
+                if let Some(target_block_number) = target_block_number {
+                    // Wait for Subspace block importing notifications
+                    let mut block_importing_notification_stream =
+                        Box::pin(params.operator_streams.block_importing_notification_stream);
+
+                    while let Some((block_number, mut acknowledgement_sender)) =
+                        block_importing_notification_stream.next().await
+                    {
+                        trace!(%block_number, "Acknowledged block import from consensus chain.");
+                        if acknowledgement_sender.send(()).await.is_err() {
+                            return Err(sp_consensus::Error::Other(
+                                format!("Can't acknowledge block import #{}", block_number).into(),
+                            ));
+                        }
+
+                        if block_number >= target_block_number.into() {
+                            break;
+                        }
+                    }
+
+                    // Drain Substrate block imported notifications
+                    let mut imported_block_notification_stream =
+                        Box::pin(params.operator_streams.imported_block_notification_stream);
+
+                    while let Some(import_notification) =
+                        imported_block_notification_stream.next().await
+                    {
+                        let block_number = *import_notification.header.number();
+                        trace!(%block_number, "Block imported from consensus chain.");
+
+                        if block_number >= target_block_number.into() {
+                            break;
+                        }
+                    }
```

Comment:
The underlying stream type is created here: 
https://github.com/autonomys/polkadot-sdk/blob/819a5818284a96a5a5bd65ce67e69bab860d4534/substrate/client/service/src/client/client.rs#L1953

It will warn on 100,000 items, so that seems ok for now, but maybe we could leave a TODO here?

Reply:
I agree that the suggested snippet is more succinct, however, IMO it's better to separate those streams because the first one impacts other parts of the system and is a source of deadlocks in two networks. It's easier to deal with it this way.